File: DataStructures/Stacks/atcoder/B51/B51-Claude.py
# PythonでカッコマッチングのコードをTypeScript版から変換して作成します。型ヒントを使用し、パフォーマンスとメモリ効率に注意を払います。Pythonでのカッコマッチングソリューションを作成しました。主な特徴：

# ## 型安全性とPylance対応
# - すべての関数に詳細な型ヒント（`typing`モジュール使用）
# - 変数の型も明示的に宣言
# - Pylanceの型チェックに対応した記述

# ## パフォーマンス最適化
# - **時間計算量**: O(n log n) - スタック処理O(n) + ソートO(n log n)
# - **空間計算量**: O(n) - スタックと結果配列の効率的使用
# - `list.append()`と`list.pop()`でスタック操作を高速化
# - リスト内包表記で文字列結合を最適化

# ## メモリ使用量監視
# - `tracemalloc`でメモリ使用量を正確に測定
# - ピークメモリと現在メモリの両方を監視
# - 理論値と実測値の比較が可能

# ## 主要関数の詳細

# ### `find_bracket_pairs(s: str) -> List[List[int]]`
# - カッコの対応を発見する核心処理
# - スタックでLIFO操作を実現
# - 1-indexedの位置で結果を返却

# ### `sort_pairs(pairs: List[List[int]]) -> List[List[int]]`
# - 出力条件 `max(li, ri) < max(li+1, ri+1)` に従ってソート
# - `sorted()`関数でラムダ式を使用した効率的ソート

# ### `measure_performance() -> None`
# - `time.perf_counter()`で高精度時間測定
# - `tracemalloc`でメモリ使用量を詳細分析
# - デバッグ情報は`sys.stderr`に出力

# ## エラーハンドリング機能
# - `validate_input()`で入力の妥当性チェック
# - カッコのバランス検証
# - 不正文字の検出

# ## デバッグ支援機能
# - `debug_solve()`で詳細な実行トレース
# - 各ステップでのスタック状態を記録
# - 開発時のデバッグに有用

# TypeScript版と同等の機能を持ちながら、Python特有の最適化（リスト内包表記、`enumerate()`等）を活用して効率的に実装しています。制約内（200,000文字、2秒、1024MiB）で確実に動作するよう設計されています。

import sys
import time
import tracemalloc
from typing import List, Tuple


def find_bracket_pairs(s: str) -> List[List[int]]:
    """
    カッコ列の対応を見つける関数
    
    Args:
        s (str): 対応の取れているカッコ列
        
    Returns:
        List[List[int]]: 対応するカッコのペアのリスト [[左の位置, 右の位置], ...]
                        位置は1-indexedで返される
    """
    stack: List[int] = []  # 開きカッコの位置を保存するスタック
    pairs: List[List[int]] = []  # 対応するカッコのペア
    
    # 各文字を順番に処理
    for i, char in enumerate(s):
        if char == '(':
            # 開きカッコの場合、位置をスタックに追加（1-indexedに変換）
            stack.append(i + 1)
        else:  # char == ')'
            # 閉じカッコの場合、対応する開きカッコをスタックから取得
            open_pos: int = stack.pop()
            close_pos: int = i + 1  # 1-indexedに変換
            pairs.append([open_pos, close_pos])
    
    return pairs


def sort_pairs(pairs: List[List[int]]) -> List[List[int]]:
    """
    カッコのペアを指定された順序でソートする関数
    条件: max(li, ri) < max(li+1, ri+1)
    
    Args:
        pairs (List[List[int]]): カッコのペアのリスト
        
    Returns:
        List[List[int]]: ソート済みのカッコのペアのリスト
    """
    # ソートのキーとして各ペアの最大値を使用
    return sorted(pairs, key=lambda pair: max(pair[0], pair[1]))


def solve(input_str: str) -> str:
    """
    メイン処理関数
    
    Args:
        input_str (str): 入力文字列（カッコ列）
        
    Returns:
        str: 結果の文字列（各行に左右のペアの位置）
    """
    s: str = input_str.strip()
    
    # カッコの対応を見つける
    pairs: List[List[int]] = find_bracket_pairs(s)
    
    # 指定された順序でソート
    sorted_pairs: List[List[int]] = sort_pairs(pairs)
    
    # 結果を文字列形式で出力
    result_lines: List[str] = [f"{pair[0]} {pair[1]}" for pair in sorted_pairs]
    return '\n'.join(result_lines)


def measure_performance() -> None:
    """
    メモリと時間の使用量を監視しながらメイン処理を実行する関数
    """
    # メモリ使用量の測定開始
    tracemalloc.start()
    
    # 実行時間の測定開始
    start_time: float = time.perf_counter()
    
    # 標準入力から読み取り
    input_data: str = sys.stdin.read()
    
    # メイン処理実行
    result: str = solve(input_data)
    
    # 実行時間の測定終了
    end_time: float = time.perf_counter()
    # execution_time: float = end_time - start_time
    _: float = end_time - start_time
    
    # メモリ使用量の測定終了
    # current_memory, peak_memory = tracemalloc.get_traced_memory()
    _, _ = tracemalloc.get_traced_memory()
    tracemalloc.stop()
    
    # 結果の出力
    print(result)
    
    # デバッグ用情報（実際の提出時はコメントアウト）
    # print(f"実行時間: {execution_time * 1000:.2f}ms", file=sys.stderr)
    # print(f"現在のメモリ使用量: {current_memory / 1024 / 1024:.2f}MB", file=sys.stderr)
    # print(f"ピークメモリ使用量: {peak_memory / 1024 / 1024:.2f}MB", file=sys.stderr)


def analyze_complexity(n: int) -> None:
    """
    計算量の理論的分析を行う関数
    
    Args:
        n (int): 入力サイズ（文字列の長さ）
    """
    # 時間計算量の分析
    stack_operations_time: str = "O(n)"  # スタックの操作
    sorting_time: str = "O(n log n)"  # ペアのソート
    overall_time: str = "O(n log n)"  # 全体の時間計算量
    
    # 空間計算量の分析
    stack_space: str = "O(n/2)"  # スタック（最悪の場合）
    pairs_space: str = "O(n/2)"  # ペア配列
    overall_space: str = "O(n)"  # 全体の空間計算量
    
    # メモリ使用量の推定（バイト単位）
    estimated_stack_memory: int = (n // 2) * 28  # int型（28バイト）のリスト
    estimated_pairs_memory: int = (n // 2) * 56  # 2要素のintリスト（56バイト）
    estimated_total_memory: int = estimated_stack_memory + estimated_pairs_memory
    
    print(f"=== 計算量分析（入力サイズn={n}）===", file=sys.stderr)
    print(f"時間計算量: {overall_time}", file=sys.stderr)
    print(f"  - スタック操作: {stack_operations_time}", file=sys.stderr)
    print(f"  - ソート処理: {sorting_time}", file=sys.stderr)
    print(f"空間計算量: {overall_space}", file=sys.stderr)
    print(f"  - スタック: {stack_space}", file=sys.stderr)
    print(f"  - ペア配列: {pairs_space}", file=sys.stderr)
    print(f"推定メモリ使用量: {estimated_total_memory / 1024 / 1024:.2f}MB", file=sys.stderr)


def validate_input(s: str) -> bool:
    """
    入力が有効なカッコ列かを検証する関数
    
    Args:
        s (str): 検証するカッコ列
        
    Returns:
        bool: 有効なカッコ列の場合True、そうでなければFalse
    """
    balance: int = 0
    
    for char in s:
        if char == '(':
            balance += 1
        elif char == ')':
            balance -= 1
            # 閉じカッコが開きカッコより多くなったら無効
            if balance < 0:
                return False
        else:
            # カッコ以外の文字が含まれていたら無効
            return False
    
    # 最終的にバランスが0でなければ無効
    return balance == 0


# デバッグ用の詳細実行関数
def debug_solve(input_str: str) -> Tuple[str, List[Tuple[int, str, List[int], str]]]:
    """
    デバッグ用の詳細な実行トレースを含む解決関数
    
    Args:
        input_str (str): 入力文字列
        
    Returns:
        Tuple[str, List[Tuple[int, str, List[int], str]]]: 
            結果文字列と実行トレース
            トレース形式: (位置, 文字, スタック状態, 処理内容)
    """
    s: str = input_str.strip()
    stack: List[int] = []
    pairs: List[List[int]] = []
    trace: List[Tuple[int, str, List[int], str]] = []
    
    for i, char in enumerate(s):
        pos: int = i + 1
        
        if char == '(':
            stack.append(pos)
            trace.append((pos, char, stack.copy(), f"位置{pos}をスタックに追加"))
        else:  # char == ')'
            open_pos: int = stack.pop()
            pairs.append([open_pos, pos])
            trace.append((pos, char, stack.copy(), f"ペア[{open_pos}, {pos}]を作成"))
    
    sorted_pairs: List[List[int]] = sort_pairs(pairs)
    result: str = '\n'.join(f"{pair[0]} {pair[1]}" for pair in sorted_pairs)
    
    return result, trace


if __name__ == "__main__":
    # パフォーマンス測定を含むメイン処理の実行
    measure_performance()
    
    # 大きな入力サイズでの計算量分析例
    # analyze_complexity(200000)  # 最大制約での分析

File: DataStructures/Stacks/atcoder/B51/B51-Claude.ts
// この問題は対応の取れているカッコ列で、どの位置のカッコ同士が対応しているかを見つける問題ですね。スタックを使って効率的に解決できます。このTypeScriptソリューションの主な特徴：

// ## アルゴリズムの説明
// 1. **スタックを使用**: 開きカッコ `(` の位置をスタックに保存
// 2. **閉じカッコ処理**: 閉じカッコ `)` に遭遇したとき、スタックから最新の開きカッコの位置を取得してペアを作成
// 3. **ソート**: 問題の条件 `max(li, ri) < max(li+1, ri+1)` に従ってペアをソート

// ## パフォーマンス最適化
// - **時間計算量**: O(n log n) - スタック処理がO(n)、ソートがO(n log n)
// - **空間計算量**: O(n) - スタックとペア配列の使用
// - **メモリ効率**: 不要な配列コピーを避け、インプレース処理を重視

// ## 関数設計
// - `findBracketPairs()`: カッコの対応を見つける核心処理
// - `sortPairs()`: 出力条件に合わせたソート
// - `solve()`: メイン処理をまとめる関数
// - `measurePerformance()`: 実行時間とメモリ使用量を測定

// 例として `(())()` の場合：
// - 1文字目 `(` → スタックに1を追加
// - 2文字目 `(` → スタックに2を追加  
// - 3文字目 `)` → スタックから2を取得、ペア[2,3]を作成
// - 4文字目 `)` → スタックから1を取得、ペア[1,4]を作成
// - 5文字目 `(` → スタックに5を追加
// - 6文字目 `)` → スタックから5を取得、ペア[5,6]を作成

// ソート後: [2,3], [1,4], [5,6] となり、期待される出力と一致します。

import * as fs from 'fs';

/**
 * カッコ列の対応を見つける関数
 * @param s - 対応の取れているカッコ列
 * @returns 対応するカッコのペアの配列 [左の位置, 右の位置]
 */
function findBracketPairs(s: string): number[][] {
    const stack: number[] = []; // 開きカッコの位置を保存するスタック
    const pairs: number[][] = []; // 対応するカッコのペア
    
    // 各文字を順番に処理
    for (let i = 0; i < s.length; i++) {
        if (s[i] === '(') {
            // 開きカッコの場合、位置をスタックに追加（1-indexedに変換）
            stack.push(i + 1);
        } else {
            // 閉じカッコの場合、対応する開きカッコをスタックから取得
            const openPos = stack.pop()!;
            const closePos = i + 1; // 1-indexedに変換
            pairs.push([openPos, closePos]);
        }
    }
    
    return pairs;
}

/**
 * カッコのペアを指定された順序でソートする関数
 * 条件: max(li, ri) < max(li+1, ri+1)
 * @param pairs - カッコのペアの配列
 * @returns ソート済みのカッコのペアの配列
 */
function sortPairs(pairs: number[][]): number[][] {
    return pairs.sort((a, b) => {
        const maxA = Math.max(a[0], a[1]);
        const maxB = Math.max(b[0], b[1]);
        return maxA - maxB;
    });
}

/**
 * メイン処理関数
 * @param input - 入力文字列（カッコ列）
 * @returns 結果の文字列
 */
function solve(input: string): string {
    const s = input.trim();
    
    // カッコの対応を見つける
    const pairs = findBracketPairs(s);
    
    // 指定された順序でソート
    const sortedPairs = sortPairs(pairs);
    
    // 結果を文字列形式で出力
    return sortedPairs.map(pair => `${pair[0]} ${pair[1]}`).join('\n');
}

// メモリと時間の使用量を監視するための関数
function measurePerformance<T>(fn: () => T): { result: T, time: number, memory: number } {
    const startTime = process.hrtime.bigint();
    const startMemory = process.memoryUsage().heapUsed;
    
    const result = fn();
    
    const endTime = process.hrtime.bigint();
    const endMemory = process.memoryUsage().heapUsed;
    
    const time = Number(endTime - startTime) / 1000000; // ミリ秒に変換
    const memory = Math.max(0, endMemory - startMemory); // バイト
    
    return { result, time, memory };
}

// 入力の読み取りと処理
const input = fs.readFileSync('/dev/stdin', 'utf8');

const { result, time, memory } = measurePerformance(() => solve(input));

console.log(result);

// デバッグ用（実際の提出時はコメントアウト）
// console.error(`実行時間: ${time.toFixed(2)}ms`);
// console.error(`メモリ使用量: ${(memory / 1024 / 1024).toFixed(2)}MB`);